numpy/core/shape_base.py

```diff
@@ -361,78 +361,54 @@ def stack(arrays, axis=0):
-
-            f_reduce([
-              f_reduce([
-                f_map(1),
-                f_map(2)
-              ]),
-              f_map(3),
-              f_map(4)
-            ]])
-
-
-        State can be passed down through the calls with `f_kwargs`,
-        to iterables of mapped items. When kwargs are passed, as in
-        ``map_reduce([[1, 2], 3, 4], **kw)``, this becomes::
-
-            kw1 = f_kwargs(**kw)
-            kw2 = f_kwargs(**kw1)
-            f_reduce([
-              f_reduce([
-                f_map(1), **kw2)
-                f_map(2,  **kw2)
-              ],      **kw1),
-              f_map(3, **kw1),
-              f_map(4, **kw1)
-            ]],     **kw)
-        """
-        def f(x, **kwargs):
-            if not self.recurse_if(x):
-                return f_map(x, **kwargs)
-            else:
-                next_kwargs = f_kwargs(**kwargs)
-                return f_reduce((
-                    f(xi, **next_kwargs)
-                    for xi in x
-                ), **kwargs)
-        return f(x, **kwargs)
-
-    def walk(self, x, index=()):
-        """
-        Iterate over x, yielding (index, value, entering), where
-
-        * ``index``: a tuple of indices up to this point
-        * ``value``: equal to ``x[index[0]][...][index[-1]]``. On the first iteration, is
-                     ``x`` itself
-        * ``entering``: bool. The result of ``recurse_if(value)``
-        """
-        do_recurse = self.recurse_if(x)
-        yield index, x, do_recurse
-
-        if not do_recurse:
-            return
-        for i, xi in enumerate(x):
-            # yield from ...
-            for v in self.walk(xi, index + (i,)):
-                yield v
+def _block_check_depths_match(arrays, index=[]):
+    def format_index(index):
+        idx_str = ''.join('[{}]'.format(i) for i in index if i is not None)
+        return 'arrays' + idx_str
+    if isinstance(arrays, tuple):
+        raise TypeError(
+            '{} is a tuple. '
+            'Only lists can be used to arrange blocks, and np.block does '
+            'not allow implicit conversion from tuple to ndarray.'.format(
+                format_index(index)
+            )
+        )
+    elif isinstance(arrays, list) and len(arrays) > 0:
+        indexes = [_block_check_depths_match(arr, index + [i])
+                   for i, arr in enumerate(arrays)]
+
+        first_index = indexes[0]
+        for i, index in enumerate(indexes):
+            if len(index) != len(first_index):
+                raise ValueError(
+                    "List depths are mismatched. First element was at depth "
+                    "{}, but there is an element at depth {} ({})".format(
+                        len(first_index),
+                        len(index),
+                        format_index(index)
+                    )
+                )
+        return first_index
+    elif isinstance(arrays, list) and len(arrays) == 0:
+        return index + [None]
```

Comment:
Can you explain what's going on here?

Reply:
If an empty list is encountered the recursion needs to back up and an error message may need to be generated, if the depths don't match.

I used the length of the `index` list for the depth info in the error message. On this line I'm making sure that the length of index indeed reflects the depth of nested lists, but using `None` to flag that this actual index shouldn't be included in the error message. This is parsed at the end of [this line](https://github.com/j-towns/numpy/blob/simplify-block-implementation/numpy/core/shape_base.py#L366). I could have used some other value (such as -1) to flag an empty list.

Reply:
As an example (from the tests), if you do
```python
np.block([1, []])
```
the error that you get is
```
ValueError: List depths are mismatched. First element was at depth 1, but there is an element at depth 2 (arrays[1])
```
At the end of the message only one index is printed even though the depth is two — that's the kind of situation I'm preparing for in the above line.